### ISSUE TITLE
liftST

### ISSUE THREAD

**Issue by wyager** [josefs/STMonadTrans#5]
Here's some code I wrote for https://hackage.haskell.org/package/FastPush

    liftST :: Applicative m => ST s a -> STT s m a
    liftST (ST f) = STT (\s -> let (# s', a #) = f s in pure (STTRet s' a))

It lets you embed things like MVector writes into STT. You may want to include it if this seems reasonably sane.

**Comment by josefs** [comment]
Thanks for the suggestion. I have quite a few thoughts about `liftST`. Bare with me.

Currently `STMonadTrans` is written using the primitive functions provided by GHC for manipulating state, such as `readMutVar#`. An unfortunate consequence of this method of implementation is that I had to define a new type `STRef` which is different from the `STRef` in the `ST`-monad, because I needed access to the underlying `MutVar#`. Hence, my library contains what is essentially duplications of types and functions on these types.

Your new function `liftST` suggests a whole different way of implementing the `STMonadTrans` library, namely to reuse all the types and functions for the `ST`-monad and use `liftST` as the means to reuse. I really like that idea, I hadn't thought of it before.  So I've imlemented just that.

The function is exported only from the `Internal` module for now.
